R/select.R

```diff
@@ -0,0 +1,298 @@
+                                                    "first columns",
+                                                    "last columns",
+                                                    "range",
+                                                    "first date-time periods",
+                                                    "last date-time periods",
+                                                    "date range"),
+                             column.selections = NULL,
+                             unit = NULL,
+                             calendar = TRUE,
+                             ...)
+{
+    row.selection.mode <- tolower(row.selection.mode)
+    column.selection.mode <- tolower(column.selection.mode)
+    row.selection.mode <- match.arg(row.selection.mode)
+    column.selection.mode <- match.arg(column.selection.mode)
+
+    table.out <- FlattenTableAndDropStatisticsIfNecessary(table)
+    if (!is.null(row.selections))
+    {
+        row.selection.mode <- tolower(row.selection.mode)
+        table.out <- selectFromRows(table.out, row.selection.mode,
+                                    row.selections, unit = unit[1],
+                                    calendar = calendar[1], ...)
+    }
+
+
+    if (NCOL(table.out) > 1L && !is.null(column.selections))
+    {
+        if (length(calendar) == 2)
+            calendar <- calendar[2]
+        if (length(unit) == 2)
+            unit <- unit[2]
+        table.out <- selectFromColumns(table.out, table.out,
+                                       selection.mode = column.selection.mode,
+                                       selections = column.selections,
+                                       unit = unit, calendar = calendar, ...)
+
+    }
+
+
+    if (!is.null(attr(table.out, "name")))
+    {
+        attr(table.out, "name.original") <- attr(table.out, "name")
+        attr(table.out, "name") <- NULL
+    }
+    return(table.out)
+}
+
+selectFromRows <- function(table, selection.mode = "vector",
+                           selections = NULL, unit, calendar, ...)
+{
+    table.out <- table
+    if(selection.mode == "range")
+        selections <- parseRangeString(selections)
+    else if (selection.mode == "date range")
+        selections <- findDatesInTable(table, selections, 1)
+    else if (selection.mode == "first rows")
+        unit <- "Row"
+
+    n.dims <- length(dim(table))
+
+    if (grepl("^first", selection.mode))
+    {
+        table.out <- First(table, keep = selections,
+                           unit = unit, calendar = calendar, ...)
+        selections <- seq_len(nrow(table.out))
+    }else if (grepl("^last", selection.mode)){
+        table.out <- Last(table, keep = selections,
+                          unit = unit, calendar = calendar, ...)
+        selections <- (nrow(table) - nrow(table.out) + 1):nrow(table)
+    }else
+    {
+        if (is.character(selections))
+            selections <- checkSuppliedIndices(table, selections, 1)
```

Comment:
Need to check that selections are within range otherwise throw nice error

Reply:
Does order of parsed selections matter or do we just sort numerically? How about duplicate selections?

Reply:
I've added a check of the numeric indices/ranges in a function `checkNumericSelections()` with an informative error/warning message if all/any selections are invalid and need to be ignored.

The order does matter. I've made this clear in the unit tests. Duplicate selections are ignored.